WIP: RFC for migrating cluster-ui into DB Console

nathanstilwell opened the pull request:
# RFC: Cluster UI Code Organization

This RFC proposes changing the architecture of code shared between DB Console and CockroachCloud Console to increase ease of development. Some current challenges in development are outlined and initial solutions are proposed as well as some recommendations for future improvements. 

View RFC [here](https://github.com/nathanstilwell/cockroach/blob/db-console-cluster-ui-rfc/docs/RFCS/20210415_db_console_cluster_ui.md)

Commit:
rfc: Draft RFC for migrating cluster-ui components into DB Console

Drafted an rfc to migrate components that were extracted for use in
CockroachCloud console back into DB Console as independant versions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64085)
<!-- Reviewable:end -->

Commit:
rfc: Additional Draft RFC for migrating cluster-ui

Updating copy and diagrams based on addional conversation. Fleshing out
draft version of the rest of the sections to communicate a way forward.

Release note: None

maryliag commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+The current architecture to share common "page" components for cluster observability across DB Console and CockroachCloud Console makes development and maintenance of those components difficult. The motivation behind moving these components out of a repository containing an application was to not show preference to a specific context; to treat each application as a generic consumer. In practice,  ...
+
+The components that are rendered for `/statements`, `/transactions`, and `/sessions` were extracted (the code copied and then refactored) from DB Console into another repo intended to publish UI dependencies. The motivation here was that they would be consumed by CockroachCloud console as a dependency. The _presentational_ components themselves had depenencies on may other presentational componen ...
```
nit: Transaction Details (typo Tranaction)

maryliag commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## Deprecate the package in UI repo
+
+As a final step, the code in `cockroachdb/ui` should be deprecated, the `package.json` set to private, and the Github Action for publishing removed. We should keep the code around for reference or back up until we are confident it is no longer needed.
```
Here you mean just the cluster-ui package or all packages on the ui repo?

nathanstilwell replied:
Oh, just the `cluster-ui` package. I'll change that to be more explicit. We will keep all the other packages in that repo.

maryliag commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## Cluster UI
+
+The Cluster UI team has the most familiarity with the current architecture. To assist in the transition of ownership of these components (to the SQL Observability team) we can commit to doing the bulk of the work mentioned in the above section. Cluster UI would take on moving the code and git commit history into the CRDB repo, updating the build systems, tools, and dependencies required, and conf ...
```
typo: updating the _depedency_    (dependency)

maryliag commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## SQL Observability
+
+As the new stewards of the observability components, it is our recommendation to the SQL Observabilty team to consider relocating code specifically engineered for state management and routing in CockroachCloud console to that repo. This should simplify the code in that is pubilshed, the dependency tree, and the code that handles external reducers and routes in that application.
```
typo: pubilshed (published)

maryliag commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## CockroachCloud Platform
+
+One issue that has been repeatedly encountered by the Cluster UI team in supporting these shared components is resolving conflicts with global styles in CockroachCloud Console. It is our recomendation that the web component paradigmn is best served by removing global css entirely from that application. CC-Console has the ability to use modular css at the component level (CSS Modules). Where style ...
```
which repo those design token will exist?

nathanstilwell replied:
The common design tokens we have currently exist [here](https://github.com/cockroachdb/ui/tree/master/packages/design-tokens).

maryliag reviewed (commented):
Proposal looks great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-:-M_2IrDrAu-r15tqpy2y:b-5ehrr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nathanstilwell)



<!-- Sent from Reviewable.io -->

Commit:
rfc: cluster-ui reorg, spellcheck

Release note: None

Commit:
rfc: cluster ui reorg

Missed a word in this sentence.

Release note: None

nathanstilwell commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## Moving the code
+
+Relocating the code from the UI repo into the `ui` package of the CRDB repo will require some adjustments to be made to the build configuration (webpack) to include and ignore portions of the code. Since much of the package is not intended for consumption in DB Console, but will still need to be published as a npm package, we want to ensure that all builds work successfully. I anticipate this wil ...
```
Yes, the recommendation is to have a `pkg/ui/cluster-ui` directory which would contain a `package.json` and would be published in the same way. You are exactly right, this would preserve the functionality of the dependency for CC Console.

nathanstilwell commented on docs/RFCS/20210415_db_console_cluster_ui.md:
```diff
@@ -0,0 +1,74 @@
+
+## Updating CC Console
+
+When we have confirmed that both DB Console can be built successfully and the Cluster UI package can be build independently, publishing for Cluster UI should be configured. Currently in the UI repo, the `cluster-ui` package is published by a Github Action when pull requests are merged to master. Eventually it would be highly beneficial to have publishing be a part of merging code, we anticipate t ...
```
I imagined that would be a follow up after the work of this RFC (to publish on merge or some other automated build process). We won't need to publish the protobuf clients any longer so I we would also deprecate that package. I should state that explicitly in the RFC.

Commit:
rfc: cluster-ui reorg

Adding some clarification based on feedback and questions.

Release note: None

laurenbarker reviewed (commented):
Overall, this makes sense to me. I haven't worked with those packages, though, so others would know better.

A nit related to formatting: 
It seems like each paragraph is on one line which makes it difficult to read. I think the norm is to wrap at either 80 or 100 characters, and most editors can be configured to auto-wrap.

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-:-M_Ltrj4DDN7OpoU_PUY:brmgkdx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @jordanlewis, @koorosh, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LtPe4CMyroTdV4Umv:-M_Ln1bi5DHqWKuJcIQU:b-3kdhtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L53)):*
> ```Markdown
> Once the move and cleanup has happened on master, an update will need to be made to `release-20.2` as this version also depends on the an external `cluster-ui` as well. The version supporting 20.2 is on a separate branch in the UI repository (`cluster-ui-20.2`) and so while the code move on master could be referenced I'm not confident it can be duplicated. Depending on the state of the release, the work on CRDB master may need to be backported to `release-21.1` as well.
> 
> One benefit of having the components in the same repository as the protobufs, is that we will no long need to publish the protobuf clients as a dependency. After the build has configured to include these clients as a local dependency, the `@cockroachlabs/crdb-protobuf-client` package can be deprecated and the `package.json` (`/pkg/ui/src/js/package.json`) can be set to private.
> ```

nit: `no long` --> `no longer`
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 61 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_Ls_pw3snc1pM72tS9:-M_Ls_pw3snc1pM72tSA:b-7e3cdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L61)):*
> ```Markdown
> ## Deprecate `cluster-ui` in UI repository
> 
> As a final step, the `cluster-ui` code in `cockroachdb/ui` (`/packages/cluster-ui`) should be deprecated, the `package.json` set to private, and the Github Action for publishing removed. We should keep the code around for reference or back up until we are confident it is no longer needed.
> ```

I would think once we are depending on the `cluster-ui` package published from `cockroachdb/cockroach`, we could safely remove the `cluster-ui` code from `cockroachdb/ui`. If for some reason we need to use the old package, we can restore it from the git history.
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 71 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LvF2C6_TWxssmqVOn:-M_LvF2C6_TWxssmqVOo:bqxkf3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L71)):*
> ```Markdown
> ## SQL Observability
> 
> As the new stewards of the observability components, it is our recommendation to the SQL Observability team to consider relocating code specifically engineered for state management and routing in CockroachCloud console to that repository. This should simplify the code in that is published, the dependency tree, and the code that handles external reducers and routes in that application.
> ```

Moving the CC specific code to the `managed-service` repo makes sense to me. It should make it easier to work with redux, react-router, etc. It should also make it easier to tell how those pages integrate with the redux store.
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LpBfe8sdnwfAtwpZZ:-M_LpBffAwMc-iPIB7gl:b-xov25p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L75)):*
> ```Markdown
> ## CockroachCloud Platform
> 
> One issue that has been repeatedly encountered by the Cluster UI team in supporting these shared components is resolving conflicts with global styles in CockroachCloud Console. It is our recommendation that the web component paradigm is best served by removing global css entirely from that application. CC-Console has the ability to use modular css at the component level (CSS Modules). Where styles need to be kept consistent across components, we suggest the use of design tokens or at the very least Sass mixins. 
> ```

I don't think removing global styles from CC Console falls within the scope of this RFC. 

The global styles in CC Console use design tokens for color values, and rely on Sass mixins, so it's not clear what the suggested change is here. I imagine the conflicts are coming from `console/assets/css/global/_base.scss`, since those apply styles to elements. If there are conflicting styles, we can certainly work to refactor them, but styles on the components should be a higher specificity than the global styles. I imagine unexpected styles are being applied rather than styles getting overridden.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
Nice RFC @nathanstilwell. Addresses all the current pain points. Looking forward to seeing these changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-:-M_Qz4NGDlBhNHkgcuIb:b-3dfbch)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster, @jordanlewis, @koorosh, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 25 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_QwnSEBGqdZV1AvE1Z:-M_QwnSEBGqdZV1AvE1_:b-5w5ukh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L25)):*
> ```Markdown
> # Proposal
> 
> This leads me to how we might simplify this situation. Since the only code we are using from `@cockroachlabs/cluster-ui` are the presentational components, by moving the refactored components back into the CRDB codebase there will be no symlinks required to develop features there. By using the CRDB repository as the codebase for the cluster-ui pages used for CockroachCloud, the version of the pages and the underlying API (protos-js) will be controlled with versioned release branches and backports. This will create a great advantage for developing in DB Console as there will be no symlinked dependencies. Local development of CockroachCloud Console will still require symlinking to the CRDB repository, but it will be reduced to a single manageable symlink.
> ```

nit: me -> us
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_QxLvB6R6d6V30QkNX:-M_QxLvB6R6d6V30QkNY:b-uqhsvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L31)):*
> ```Markdown
> ![CC-Console dependency](http://www.plantuml.com/plantuml/png/VP31IWCn48RlUOfXJ_OGzYuYrHL1KUd5Kzj3DfbjOvDCI2O5aRwxiOMwgz1R8F_CztyppwmI5-y4MtO8JUTAPapm_WG68mQ3GN-23T1rq578L4DNG-6ISDQ8gFHERUhOm_tlhoQzIwJqrlJj-Tt6tQOjTjR0brZg4qGMknRBc8Wfou-y5WSHY15wOYOFDC0u2TSLOH26H7uL9f1pQ1KyxT705XEA8IbVgvsJJlMvzIhKmaPcJ7khyLYdgZqc2bnjuIx6dnxQKJ7Gl2rUuLdjLTR9HeR5IJSVf-RJnMEdWnXhWRpzA_92MhdHWVztW1ifuH9Zib5Mn8Lm9IvAxspUeW4c2D6BUIKLxxu1)
> 
> This also has the advantage of reducing the discussion of multiple versions to largely a discussion of the version of the database. DB Console will have no consideration of the version of Cluster UI as it will simply be the one in the codebase. Backports will be managed in the same way as other code changes in DB Console, using git branches.
> ```

nit: be explicit here instead of "this".
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_QxXWs7VoSVL2R4Ozs:-M_QxXWs7VoSVL2R4Ozt:b-7phr3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a36762d05020be73e9f965f866065455b561d5f/docs/RFCS/20210415_db_console_cluster_ui.md#L35)):*
> ```Markdown
> ![release branches](http://www.plantuml.com/plantuml/png/VT313e8m3CRn-vwYuJuK7i3m7eH3ALMKmKWxJiPtTxf8ucAyBVzy-MLfem4bRBE3eRVa5_MEhR1ZLBQzu48Zgp5dmPcX84-JUAHnw0_xku2x0LYH9hp4JJkZ1fOkPMXNyS6hlIau3AoX0tk2bjpsMjdjkkUuVfWlwVEFkYgh7zuxvrgA_x1ZMISiqsBFjHGJ5hc6bo6UoZm1)
> 
> And in support of multiple cluster versions in CockroachCloud, the published package (`@cockroachlabs/cluster-ui`) could be aligned to the version of the DB for convenience.
> ```

Clarify what you mean by "aligned to the version". I assume it's referring to the fact that published npm versions will match DB numbers.


<!-- Sent from Reviewable.io -->

Commit:
rfc: Cluster UI migration

Adjusting typos and language based on feedback.

Release note: None

nathanstilwell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster, @jordanlewis, @koorosh, @laurenbarker, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MYuWTpwIfpkBM_QsNca-r2-59:-M_RL7-d0TlZzOqILETm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a80ae7c3eca689feb1e1b9f900276dfee2aff/docs/RFCS/20210415_db_console_cluster_ui.md#L59)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

Oh, just the `cluster-ui` package. I'll change that to be more explicit. We will keep all the other packages in that repo.
</blockquote></details>

Done.
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MYuWTpwIfpkBM_QsNca-r2-65:-M_RKonc4jnFINQINIyn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a80ae7c3eca689feb1e1b9f900276dfee2aff/docs/RFCS/20210415_db_console_cluster_ui.md#L65)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

typo: updating the _depedency_    (dependency)
</blockquote></details>

Done.
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MYuWTpwIfpkBM_QsNca-r2-69:-M_RKr_d7yEH7A8OjgYj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a80ae7c3eca689feb1e1b9f900276dfee2aff/docs/RFCS/20210415_db_console_cluster_ui.md#L69)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

typo: pubilshed (published)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nathanstilwell commented:
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_QxXWs7VoSVL2R4Ozs:-M_RLsCC9XXreniYkP6d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aef872fa9017d2d1cf6c56f9421449fc5ad41147/docs/RFCS/20210415_db_console_cluster_ui.md#L35)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Clarify what you mean by "aligned to the version". I assume it's referring to the fact that published npm versions will match DB numbers.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nathanstilwell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @jordanlewis, @koorosh, @laurenbarker, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LpBfe8sdnwfAtwpZZ:-M_RdcjG6kySSaCJDO9L:b4m3t6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/aef872fa9017d2d1cf6c56f9421449fc5ad41147/docs/RFCS/20210415_db_console_cluster_ui.md#L75)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

I don't think removing global styles from CC Console falls within the scope of this RFC. 

The global styles in CC Console use design tokens for color values, and rely on Sass mixins, so it's not clear what the suggested change is here. I imagine the conflicts are coming from `console/assets/css/global/_base.scss`, since those apply styles to elements. If there are conflicting styles, we can certainly work to refactor them, but styles on the components should be a higher specificity than the global styles. I imagine unexpected styles are being applied rather than styles getting overridden.
</blockquote></details>

The recommendation to remove global styles is purely a recommendation. I suppose I could have been a little more verbose in why I believe that having global styles in an application built from components is harmful. The promise of web components was one of modularity and controlled scope. Global styles disrupt that practice. At best they create the need for conventions to control specificity in the cascade. At worst they create unexpected visual bugs.

I suggested design tokens and possible mixins because these are techniques we already have in place to promote consistency. I was suggesting that we continue to use these techniques rather than have broad global styles applied. You are correct that most often these conflicts were encountered from global type selectors (in `console/assets/css/global/_base.scss` and also in `antd/dist/antd.css`). The styles on the components are indeed higher specificity (or can be made to be higher), but then we have a component then needing to have an override for a specific application context.

I will admit that this is an aspirational recommendation rather than a practical one. I am certainly not saying that a deep refactoring of CSS is required for this RFC, but I wanted to raise this as an issue that we have seen and I anticipate coming up again.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @jordanlewis, @laurenbarker, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_UQE9mEWFCUG-gzgmM:-M_UQE9mEWFCUG-gzgmN:b-si2i54) (no related file):*
@nathanstilwell , would it be appropriate to describe here the future usage of `cluster-ui` package to answer following questions:
- when new features should be added in `cluster-ui` package?
- should we consider `cluster-ui` package as a place for shared code (utility functions, etc) or it's only for page components that are shared with CC?

---


<!-- Sent from Reviewable.io -->

nathanstilwell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @jordanlewis, @koorosh, @laurenbarker, @maryliag, @nathanstilwell, @nkodali, and @petermattis)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_UQE9mEWFCUG-gzgmM:-M_VtvLE59i2eHDHpRng:bxc1gd1) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@nathanstilwell , would it be appropriate to describe here the future usage of `cluster-ui` package to answer following questions:
- when new features should be added in `cluster-ui` package?
- should we consider `cluster-ui` package as a place for shared code (utility functions, etc) or it's only for page components that are shared with CC?
</blockquote></details>

I think it would. While I think ultimately the answer to those questions will be up to the SQL Observability team (or any future team owning specific components shared between DB Console and CC Console), I think I can represent the recommendation of Cluster UI in this regard.

---
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LpBfe8sdnwfAtwpZZ:-M_VuXT21QC5MTv_wudt:b8up8jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aef872fa9017d2d1cf6c56f9421449fc5ad41147/docs/RFCS/20210415_db_console_cluster_ui.md#L75)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

@nathanstilwell, if it's causing UI bugs, I would suggest creating a Jira ticket for the CC Console team with examples of where global styles are causing unexpected behavior for `cluster-ui` components. Changing the global styles could impact many different pages of CC Console, which is why it seems out-of-scope for engineers focused on relocating the `cluster-ui` package.
</blockquote></details>

This is a good suggestion and in hindsight it would have been helpful for us to increase the visibility of these issues as we encountered them. The Cluster UI team will work outside of this RFC to provide some documentation around issues historic and ongoing and share it with the CockroachCloud team.


<!-- Sent from Reviewable.io -->

Commit:
rfc: Cluster UI code reorg, preferred places to share code

Offered an opinion about where to share certain types of code could be
shared between applications (`cluster-ui` package vs `ui-components`
package).

Release note: None

laurenbarker commented:
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-M_LpBfe8sdnwfAtwpZZ:-M_gPAGkBRHoAR2095gf:b-k6rpao) ([raw file](https://github.com/cockroachdb/cockroach/blob/aef872fa9017d2d1cf6c56f9421449fc5ad41147/docs/RFCS/20210415_db_console_cluster_ui.md#L75)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

This is a good suggestion and in hindsight it would have been helpful for us to increase the visibility of these issues as we encountered them. The Cluster UI team will work outside of this RFC to provide some documentation around issues historic and ongoing and share it with the CockroachCloud team.
</blockquote></details>

Sounds good, thanks!


<!-- Sent from Reviewable.io -->

Commit:
rfc: Cluster UI RFC

Changed some language around how the dependency on cluster ui will be
declared, after some prototyping.

Release note: None

dhartunian reviewed (commented):
Just took a second pass. Only style/grammar nits from me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-:-MaTlfsd17OGeKTJG15N:bxtw2cx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster, @jordanlewis, @koorosh, @laurenbarker, @maryliag, @nathanstilwell, and @petermattis)

___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 12 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTiPqF2Dj8ELPoVaW9:-MaTiPqF2Dj8ELPoVaWA:bwy5nfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L12)):*
> ```Markdown
> The current architecture to share common "page" components for cluster observability across DB Console and CockroachCloud Console makes development and maintenance of those components difficult. The motivation behind moving these components out of a repository containing an application was to not show preference to a specific context; to treat each application as a generic consumer. In practice, this simply isn't true. CockroachDB is the API that drives these components (through protocol buffers) and the majority of the work to create the collection of components known as "Cluster UI" was to mirror these views in CockroachCloud. Sharing these components has been successful and the "Cluster UI" components function in CockroachCloud, but developing these components for DB Console has become more arduous.
> 
> The components that are rendered for `/statements`, `/transactions`, and `/sessions` were extracted (the code copied and then refactored) from DB Console into another repository intended to publish UI dependencies. The motivation here was that they would be consumed by CockroachCloud console as a dependency. The _presentational_ components themselves had dependencies on may other presentational components that also needed to be copied. These components also depended on _connected_ components to take relevant data from the application state and pass them as props to the component. These components are generally lists of things that _link_ to another component to display an individual item (Statements -> Statement Details, Transactions -> Transaction Details, etc) and so also depended on the ability to use the applications routing. All of these mechanisms were established in DB-Console, but are incompatible with the CockroachCloud console application. So code had to be created to integrate these features into the application state and routes of CC Console.
> ```

nit: ...dependencies on *many* other..., ...*application's* routing...
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 12 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTibi74Sqaa_UPsUqe:-MaTibi74Sqaa_UPsUqf:b-ng7pp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L12)):*
> ```Markdown
> The current architecture to share common "page" components for cluster observability across DB Console and CockroachCloud Console makes development and maintenance of those components difficult. The motivation behind moving these components out of a repository containing an application was to not show preference to a specific context; to treat each application as a generic consumer. In practice, this simply isn't true. CockroachDB is the API that drives these components (through protocol buffers) and the majority of the work to create the collection of components known as "Cluster UI" was to mirror these views in CockroachCloud. Sharing these components has been successful and the "Cluster UI" components function in CockroachCloud, but developing these components for DB Console has become more arduous.
> 
> The components that are rendered for `/statements`, `/transactions`, and `/sessions` were extracted (the code copied and then refactored) from DB Console into another repository intended to publish UI dependencies. The motivation here was that they would be consumed by CockroachCloud console as a dependency. The _presentational_ components themselves had dependencies on may other presentational components that also needed to be copied. These components also depended on _connected_ components to take relevant data from the application state and pass them as props to the component. These components are generally lists of things that _link_ to another component to display an individual item (Statements -> Statement Details, Transactions -> Transaction Details, etc) and so also depended on the ability to use the applications routing. All of these mechanisms were established in DB-Console, but are incompatible with the CockroachCloud console application. So code had to be created to integrate these features into the application state and routes of CC Console.
> ```

nit: standardize the naming of of "DB Console" and "CC Console". Capitalization and use of hyphen is a bit inconsistent here
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 18 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTjgWA12kElivM15j7:-MaTjgWA12kElivM15j8:b-894y34) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L18)):*
> ```Markdown
> ![cluster ui dependency](http://www.plantuml.com/plantuml/png/VP91Yzim48Nl_XM3Jzs3y3sKihlUImjRba9FcnnaQUoeaMQ4D0g4qlzUEqvhXwIzsUXxpzyJVioYavJWAt7Y4UhMbooOuFjdi1YHB98vWIDuUGPD5jeMMgRC8_l1b9IGvA6keZO7FOpnaLgEsHmIQxEtNVFtkOc7eIdkeOQVrVkAVefnLxt7nC9P_SYxwbT5B0WTYt00xc5r21jelKEIivAS8kOj3KYOjA25Fd7OqenFwkstb-T5SLbuhLXSSs5oOsP6_H5Tti6mnSckJpfCRUsTmLvtkGcQnYwhCgQZCpWNZVXwyTXZF1SBzTUtL_YYZpgRvAB0syLa_dgodJhFwKglp7dvRYqrzQolHts2-o7OmAkcehq_GIoTTtcFzdOUgB5XtR-1fo8Sj3XpdTqI4mHC0u8m13q5xlHzLEeVHZTNChFhRTDlyd_Y4ScMu7y3)
> 
> For local development, dependency management tools can create symlinks for dependency development. Developing with a single dependency, and thus a single symlink, is fairly trivial and a common practice among developers of published npm packages. Adding layers of dependency, and thus symlinks, increases the difficulty and combining layers of dependency and crisscrossing repositories increases it further. This is due to the nature of frontend development tools needing to "watch and rebuild" these symlinked dependencies and so the more symlinks your are managing, the more build processes you are juggling simultaneously. From the diagram above you can see that the `cluster-ui` package continues to depend on the `protos-js` in DB Console that need to be published to npm to be consumed as a dependency. For the purposes of local development in DB Console, this means a change to the _backend_ and the _frontend_ for one of these "pages" would require the engineer to create two symlinks and potentially manually rebuild two or three times to see an update in the UI. This is a significant hindrance. This hindrance becomes more exacerbated when considering multiple DB versions.
> ```

Minor: consider mentioning that this increases number of PRs needed (and in a specific order!) to get a feature merged
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTki5pBt9LRzMFrGJN:-MaTki5pBt9LRzMFrGJO:b-htvmjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L35)):*
> ```Markdown
> ![release branches](http://www.plantuml.com/plantuml/png/VT313e8m3CRn-vwYuJuK7i3m7eH3ALMKmKWxJiPtTxf8ucAyBVzy-MLfem4bRBE3eRVa5_MEhR1ZLBQzu48Zgp5dmPcX84-JUAHnw0_xku2x0LYH9hp4JJkZ1fOkPMXNyS6hlIau3AoX0tk2bjpsMjdjkkUuVfWlwVEFkYgh7zuxvrgA_x1ZMISiqsBFjHGJ5hc6bo6UoZm1)
> 
> And in support of multiple cluster versions in CockroachCloud, the published package (`@cockroachlabs/cluster-ui`) should be converted to [CalVer](https://calver.org/) versions (instead of there current [Semver](https://semver.org/) versions) based on the version of DB Console.
> ```

BTW: not sure how the version will be set in npm here but would be nice for Cluster UI to automagically inherit the same version as the DB does on build.
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 50 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTkW9W-DcIjLIFw63C:-MaTkW9W-DcIjLIFw63D:bygwznx) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L50)):*
> ```Markdown
> 
> 
> After the code is moved, dependencies have been resolved, and the build is successful the dependency that references the `@cockroachlabs/cluster-ui` dependency should be updated to point to local code instead. This will use npm syntax to reference a local directory without a version, making it possible to leave dependency references in place in code but achieve a better local development experience. It was originally expected to refactor DB Console code to reference Cluster ui as a local directory instead of an external dependency, but this turned out to create untenable build issues for DB Console. Continuing to reference cluster ui as an external dependency (when in reality it is a local directory) allow for better ergonomics when building both DB Console and the cluster ui package independently for publishing.
> ```

nit: standardize capitalization of "Cluster UI" to match rest of doc
___
*[docs/RFCS/20210415_db_console_cluster_ui.md, line 95 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-MaTlPpA-2w0RIejE9w0:-MaTlPpA-2w0RIejE9w1:b-k1jsoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e1bca2075afb61369941e81b6de222405ae9fc/docs/RFCS/20210415_db_console_cluster_ui.md#L95)):*
> ```Markdown
> ## SQL Observability
> 
> As the new stewards of the observability components, it is our recommendation to the SQL Observability team to consider relocating code specifically engineered for state management and routing in CockroachCloud console to that repository. This should simplify the code in that is published, the dependency tree, and the code that handles external reducers and routes in that application.
> ```

nit: grammar: "...code in that is published..."


<!-- Sent from Reviewable.io -->

Commit:
docs: cluster-ui rfc is in progress
Release note: None

maryliag reviewed (approved):
Awesome work on this!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64085#-:-MbHlPlq9EDtjVlu2yTK:bulpmt5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @elkmaster, @jordanlewis, @koorosh, @laurenbarker, @maryliag, @nathanstilwell, and @petermattis)



<!-- Sent from Reviewable.io -->

nathanstilwell commented:
bors r+
